lib.nix

```diff
@@ -2,31 +2,115 @@
+  fetchGitPin = name: pinJ:
+    builtins.fetchGit (pinJ // { name = name; });
+
+  ## repoSpec                = RepoSpec { name :: String, subdir :: FilePath, src :: Drv }
+  ## fetchGitWithSubmodules :: Name -> Drv -> Map String RepoSpec -> Drv
+  fetchGitWithSubmodules = mainName: mainRev: mainSrc: subRepos:
+    with builtins; with pkgs;
+    let subRepoCmd = repo: ''
+        chmod -R u+w $(dirname $out/${repo.subdir})
+        rmdir $out/${repo.subdir}
+        cp -R  ${repo.src} $out/${repo.subdir}
+        '';
+        cmd = ''
+
+        cp -R ${mainSrc} $out
+
+        '' + concatStringsSep "\n" (map subRepoCmd (attrValues subRepos));
+    in runCommand "fetchGit-composite-src-${mainName}-${mainRev}" { buildInputs = []; } cmd;
+
+  fetchGitPinWithSubmodules = pinRoot: name: { submodules ? {}, ... }@pin:
+    let fetchSubmodule = subName: subDir: { subdir = subDir; src = pkgs.fetchgit (getPinFetchgit pinRoot subName); };
+    in fetchGitWithSubmodules name pin.rev
+       (builtins.fetchGit (removeAttrs pin ["submodules"]))
+       (lib.mapAttrs fetchSubmodule submodules);
+
+  ## Depending on whether the repo is private (URL has 'git@github' in it), we need to use fetchGit*
+  fetchPinAuto = pinRoot: name:
+    if pinIsPrivate pinRoot name
+    then fetchGitPinWithSubmodules pinRoot name (getPinFetchGit pinRoot name)
+    else pkgs.fetchgit                          (getPinFetchgit pinRoot name);
 
   # nixpkgs can be overridden for debugging purposes by setting
   # NIX_PATH=custom_nixpkgs=/path/to/nixpkgs
   pkgs = iohkNix.pkgs;
   lib = pkgs.lib;
+  fetchProjectPackages = name: host: pinRoot: revOverride: args:
+    let
+      src = let try = builtins.tryEval host;
+        in if try.success
+           then builtins.trace "using search host <${name}>" try.value
+           else fetchPinAuto pinRoot name;
+      src-phase2 = let
+        localOverride = {
+          ### XXX: not really workable right now, for obvious reasons.  Left for uniformity with CSL definition above/future refactoring.
+          outPath = builtins.fetchTarball "https://github.com/input-output-hk/${name}/archive/${revOverride}.tar.gz";
+          rev = revOverride;
+        };
+        in if (revOverride != null) then localOverride else src;
+      pkgs = import src-phase2 ({
+          inherit (args) enableDebugging enableProfiling;
+        } // lib.optionalAttrs (src-phase2 ? rev) {
+          gitrev = src-phase2.rev;
+        });
+    in pkgs;
+  javaOverrideNixpkgsConfig = {
+    overlays = [ oracleJdkOverlay ];
+    config.allowUnfree = true;
+  };
+  oracleJdkOverlay =  self : super: {
+    oraclejdk8 = super.callPackage ./goguen/jdk-override/jdk8cpu-linux.nix {
+      installjdk = true;
+      pluginSupport = false;
+      licenseAccepted = true;
+    };
+  };
+  graalvm8 = (import iohkNixGoguen.nixpkgs javaOverrideNixpkgsConfig).graalvm8;
 in lib // (rec {
+  inherit (iohkNix) nixpkgs;
+  inherit mkIohkNix fetchProjectPackages pkgs graalvm8;
```

Comment:
Ideally, we wouldn't want to define app-specific things in the common path, but instead downstream.

Maybe not for now, since we want to expedite this PR.